docs/testing.md

```diff
@@ -0,0 +1,81 @@
+
+## General guidelines
+
+- Imports should be alphabetized, even in test files.
+- Comments should be full sentences to ease readability.
+- There are a lot of helpers in the `tests/unit/helpers` and `tests/unit/amo/helpers` modules–please use them.
+- Use action creators in `amo/actions/` and `core/actions/` instead of hard-coding `dispatch()` arguments or state data for tests. This applies to both UI components and reducers/sagas.
+- Use constants (see `core/constants`) when using the same value across files. This avoids hard-coding values and [magic constants](https://en.wikipedia.org/wiki/Magic_constant).
+
+
+## Jest
+
+[Jest](https://facebook.github.io/jest/docs/en/getting-started.html) is our main testing framework. Please refer to the [`README` section about running the test suite](https://github.com/mozilla/addons-frontend#running-tests) to know how to run the tests. Below are a few rules regarding Jest:
+
+- Prefer `toEqual()` over `toBe()`.
+- Do not use `expect.assertions(N)`, this is hard to maintain. Instead, add a `catch()` branch to detect unexpected errors.
+
+When creating a new test file, start with a `describe()` block that takes the current file name as first argument. This makes easy to find/edit a failing test case as Jest will display the test file in its output:
+
+```js
+describe(__filename, () => {
+});
+```
+
+## Spies/Stubs/Mocks and Sinon.JS
+
+We use [sinon](http://sinonjs.org/) for spies, stubs and mocks. In addition, we use [sinon assertions](http://sinonjs.org/releases/v3.2.1/assertions/) over Jest expectations because failure messages are more descriptive.
+
+```js
+const spy = sinon.spy();
+// ...
+
+// NOT GOOD
+expect(spy.called).toEqual(true);
+
+// GOOD
+sinon.assert.called(spy);
+```
+
+There is no need to import sinon, it is already in the global scope for all test files.
+
+
+## Testing reducers and sagas
+
+For sagas/reducers, there are two useful helpers: `dispatchClientMetadata()` and `dispatchSignInActions()` (`tests/unit/amo/helpers`) that should be used to initialize state in a realistic manner. The former is used to obtain a non-authenticated state while the latter returns an authenticated state.
+
+When you need a `errorHandler` or a `errorHandlerId`, use the `createStubErrorHandler()` helper from `tests/unit/helpers`.
+
+When asserting for exceptions/errors, do not use ES6 shorthand functions/implicit return functions. It explicits what is expected to throw an error:
+
+```js
+expect(() => {
+  methodThatThrowsAnError();
+}).toThrow(/expected error message/);
+```
+
+When testing sagas, use an action creator to construct the expected actions that should be called by the saga under test:
+
+```js
+const expectedLoadAction = autocompleteLoad(results);
+
+await sagaTester.waitFor(expectedLoadAction.type);
+mockApi.verify();
+
+const loadAction = sagaTester.getCalledActions()[2];
+expect(loadAction).toEqual(expectedLoadAction);
+```
+
+## Testing UI components
```

Comment:
It might be good to show an example of using `shallowUntilTarget()` in this section.